core/checksums/src/main/java/software/amazon/awssdk/checksums/internal/DigestAlgorithm.java

```diff
@@ -27,41 +31,92 @@ public enum DigestAlgorithm {
-        digestReference = new DigestThreadLocal(algorithmName);
     }
 
     public String getAlgorithmName() {
         return algorithmName;
     }
 
     /**
-     * Returns the thread local reference for the {@link MessageDigest} algorithm
+     * Returns a {@link CloseableMessageDigest} to use for this algorithm.
      */
-    public MessageDigest getDigest() {
-        MessageDigest digest = digestReference.get();
-        digest.reset();
-        return digest;
+    public CloseableMessageDigest getDigest() {
+        MessageDigest digest = digestCache.pollFirst();
+        if (digest != null) {
+            digest.reset();
+            return new CloseableMessageDigest(digest);
+        }
+        return new CloseableMessageDigest(newDigest());
+    }
+
+    private MessageDigest newDigest() {
+        try {
+            return MessageDigest.getInstance(algorithmName);
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("Unable to fetch message digest instance for Algorithm "
+                                       + algorithmName + ": " + e.getMessage(), e);
+        }
     }
 
-    private static class DigestThreadLocal extends ThreadLocal<MessageDigest> {
-        private final String algorithmName;
+    public final class CloseableMessageDigest implements SdkAutoCloseable, Cloneable {
+        private final Supplier<MessageDigest> closedDigest = () -> {
+            throw new IllegalStateException("This message digest is closed.");
+        };
+        private Supplier<MessageDigest> digest;
+        private byte[] messageDigest;
+
+        private CloseableMessageDigest(MessageDigest digest) {
+            this.digest = () -> digest;
+        }
+
+        /**
+         * Retrieve the message digest instance.
+         */
+        public MessageDigest messageDigest() {
+            return digest.get();
+        }
 
-        DigestThreadLocal(String algorithmName) {
-            this.algorithmName = algorithmName;
+        /**
+         * Retrieve the message digest bytes. This will close the message digest when invoked. This is because the underlying
+         * message digest is reset on read, and we'd rather fail future interactions with the digest than act on the wrong data.
+         */
+        public byte[] digest() {
+            if (messageDigest != null) {
+                return messageDigest;
+            }
+            messageDigest = messageDigest().digest();
+            return messageDigest;
+        }
+
+        /**
+         * Release this message digest back to the cache. Once released, you must not use the digest anymore.
+         */
+        @Override
+        public void close() {
+            if (digest == closedDigest) {
+                return;
+            }
+
+            // Avoid over-caching after large traffic bursts. The maximum chosen here is arbitrary. It's also not strictly
+            // enforced, since these statements aren't synchronized.
```

Comment:
Would it be possible to have a "prefilled" cache of lazy loaded digests, and then when we can't use a cached digest, we have a different instance that just closes and doesn't need to interact with the cache?

Reply:
That's an option. It might be trickier to decide on the size of the cache, and then it means that not releasing a message digest to the cache will have long-term performance implications if it's one of those special "cached" digests. The advantage to this implementation is that the odd error that could fail to release the digest back to the cache doesn't really hurt.